src/v/raft/seq_stm.cc

```diff
@@ -0,0 +1,337 @@
+                      kafka::error_code::out_of_order_sequence_number);
+                }
+            }
+            return _c->replicate(_insync_term, std::move(r), opts)
+              .then([](result<replicate_result> r) {
+                  if (r) {
+                      return checked<replicate_result, kafka::error_code>(
+                        r.value());
+                  } else {
+                      return checked<replicate_result, kafka::error_code>(
+                        kafka::error_code::unknown_server_error);
+                  }
+              });
+        }
+    }
+    return _c->replicate(std::move(r), std::move(opts))
+      .then([](result<replicate_result> r) {
+          if (r) {
+              return checked<replicate_result, kafka::error_code>(r.value());
+          } else {
+              return checked<replicate_result, kafka::error_code>(
+                kafka::error_code::unknown_server_error);
+          }
+      });
+}
+
+ss::future<>
+seq_stm::catchup(model::term_id last_term, model::offset last_offset) {
+    vlog(_log.trace, "seq_stm is catching up with offset:{}", last_offset);
+    return wait(last_offset, model::no_timeout)
+      .then([this, last_offset, last_term]() {
+          vlog(_log.trace, "seq_stm catched up with offset:{}", last_offset);
+          if (last_term <= _c->term()) {
+              _insync_term = last_term;
+          }
+      });
+}
+
+void seq_stm::compact_snapshot() {
+    auto cutoff_timestamp
+      = time_since_epoch()
+        - _config.transactional_id_exiration_ms.value().count();
+    if (_oldest_session.value() < cutoff_timestamp) {
+        auto next_oldest_session = model::timestamp::now();
+        for (auto it = _seq_table.cbegin(); it != _seq_table.cend();) {
+            if (it->second.last_write_timestamp < cutoff_timestamp) {
+                _seq_table.erase(it++);
+            } else {
+                next_oldest_session = std::min(
+                  next_oldest_session,
+                  model::timestamp(it->second.last_write_timestamp));
+                ++it;
+            }
+        }
+        _oldest_session = next_oldest_session;
+    }
+}
+
+ss::future<> seq_stm::apply(model::record_batch b) {
+    const auto& hdr = b.header();
+
+    if (hdr.type == model::record_batch_type(1) && hdr.producer_id >= 0) {
+        auto pid = producer_identity{
+          .id = hdr.producer_id, .epoch = hdr.producer_epoch};
+
+        auto pid_seq = _seq_table.find(pid);
+        if (pid_seq == _seq_table.end()) {
+            seq_entry entry{
+              .pid = pid,
+              .seq = hdr.base_sequence + hdr.record_count - 1,
+              .last_write_timestamp = hdr.max_timestamp.value()};
+            _seq_table.emplace(pid, entry);
+            _oldest_session = std::min(
+              _oldest_session, model::timestamp(entry.last_write_timestamp));
+        } else if (
+          pid_seq->second.seq < hdr.base_sequence + hdr.record_count - 1) {
```

Comment:
extract `hdr.record_count -1` into a const variable in the method since you use it everywhere here.